### ISSUE TITLE
RDS Proxy Connection Error with Command-Line Options in SQL Lab

### ISSUE THREAD

**Issue by aniruddhaparte** [apache/superset#24684]
Description:
When creating a new Database connection with PSSQL and using RDS Proxy as the hostname to connect to the PSSQL server, the connection is established successfully. However, an error occurs when executing queries in SQL Lab. The error message received is "(psycopg2.Operational Error) connection to server at 'rds_proxy_endpoint', port 5432 failed: FATAL: Feature not supported: RDS Proxy currently doesn't support command-line options."

#### How to reproduce the bug

1. Install Superset version X.X.X on the operating system YYY.
2. Configure a new Database connection with PSSQL in Superset.
3. Set the hostname of the connection to the RDS Proxy endpoint: "rds_proxy_endpoint".
4. Use the appropriate port number (usually 5432) for the RDS Proxy.
5. Execute a query in SQL Lab using the newly created connection.
6. Observe the error message "(psycopg2.OperationalError) connection to server at 'rds_proxy_endpoint' , port 5432 failed: FATAL: Feature not supported: RDS Proxy currently doesn't support command-line options."

### Expected results

The query should be executed successfully without any error, using the RDS Proxy connection.

### Actual results

An error occurs with the message "(psycopg2.OperationalError) connection to server at 'rds_proxy_endpoint', port 5432 failed: FATAL: Feature not supported: RDS Proxy currently doesn't support command-line options."

#### Screenshots

![image](https://github.com/apache/superset/assets/67677161/5170c722-c390-41a5-907a-ed5142eb213d)

### Environment

(please complete the following information):

- browser type and version: Chrome Version 114.0.5735.134
- superset version: `3.0.0`
- python version: `Python 3.9.17`
- node.js version: `node -v`
- any feature flags active: "ALERT_REPORTS": True,
    "DASHBOARD_CROSS_FILTERS": True,
    "DASHBOARD_RBAC": True,
    "GENERIC_CHART_AXES": True,
    "ALLOW_FULL_CSV_EXPORT": True,
    "DRILL_TO_DETAIL": True,
    "HORIZONTAL_FILTER_BAR": True

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [ ] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [ ] I have reproduced the issue with at least the latest released version of superset.
- [ ] I have checked the issue tracker for the same issue and I haven't found one similar.

**Comment by michael-s-molina** [comment]
@aniruddhaparte What's the command line option that is being passed? Can you check [this](https://stackoverflow.com/questions/67242059/aws-rds-proxy-error-postgres-rds-proxy-currently-doesn-t-support-command-lin) and see if it resolves your problem?

**Comment by michael-s-molina** [comment]
@betodealmeida As the author of https://github.com/apache/superset/pull/23401, could take a look?

**Comment by betodealmeida** [comment]
Is @hughhhh's PR in 3.0.0? Seems like we need to cherry it to fix this problem.

**Comment by michael-s-molina** [comment]
> Is @hughhhh's PR in 3.0.0? Seems like we need to cherry it to fix this problem.

@betodealmeida https://github.com/apache/superset/pull/23868 is in 3.0. Are you talking about another PR?

**Comment by betodealmeida** [comment]
@michael-s-molina yeah, I just saw. Sorry, I always confuse RDS with Redshift.

Let me work on a fix.

**Comment by betodealmeida** [comment]
Ugh. RDS should have it's own SQLAlchemy dialect that is a thin layer over the Postgres one, instead of reusing the Postgres dialect if they're not going to be compatible. Not sure how to fix this, as we need to set the options for security reasons.

**Comment by betodealmeida** [comment]
A summary of the problem:

- Before [my fix](https://github.com/apache/superset/blob/3.0.0rc1/superset/db_engine_specs/postgres.py#L268), when using Postgres or Postgres-like databases (like RDS), in SQL Lab when you selected a schema from the dropdown **the query would not run in that schema**. Because of this, Superset was unable to evaluate if the user had permission to query an unqualified table, eg, `SELECT * FROM my_table`, since it can't know in which schema the table lives. With my fix, the query runs in the selected schema, so any unqualified table name has to be from that schema (it also ensures that there's only one schema in `search_path`, to ensure the default schema is unique).
- The way we do this is by passing `options=-csearch_path=my_schema` to the Postgres SQLAlchemy dialect when running a query. Here, `my_schema` comes from the schema dropdown in SQL Lab.
- RDS uses the same Postgres SQLAlchemy dialect, but it doesn't accept `options`. The only way to set the default path is by running a query `set search_path=my_schema` when the session starts, but currently Superset has no workflow for that, as far as I know.

In summary, I'm not sure that we can fix this for 3.0 in a secure way. To make RDS secure we need to implement a way of running pre-session queries, so that we can set the search path. I'm happy to do that refactoring, but I think cherry-picking it for 3.0 might bring in more risks.

**Comment by betodealmeida** [comment]
By the way, @mdeshmu, thanks for the detailed report, this is awesome! I'll do my best to get your queries working back again.

**Comment by betodealmeida** [comment]
This could be a fix, but I still need to test it at least with Postgres and RDS: https://github.com/apache/superset/pull/24739/files

**Comment by mdeshmu** [comment]
Thank you so much @betodealmeida for your effort.

**Comment by michael-s-molina** [comment]
Thank you @betodealmeida for the detailed description of the problem. It was really helpful to understand the root cause and it will probably be useful as the description for the PR that will fix it.

> I'm happy to do that refactoring, but I think cherry-picking it for 3.0 might be too risky.

We're still evaluating RC1. RC2 will probably be submitted next week. If you're able to submit a fix before the release, great, if not, we can always cherry pick it later in a patch release. By the way, we're planning to increase the number and cadence of patch releases for 3.0.

**Comment by mdeshmu** [comment]
fixed by #24739
